editor/src/clj/editor/model_scene.clj

```diff
@@ -143,27 +137,23 @@
                                                          (.normalize world-normal) ; need to normalize since since normal-transform may be scaled
                                                          (.get world-normal d3))))
                           normals)]
+
     ;; Raw access to run as fast as possible
     ;; 3x faster than using generated *-put! function
     ;; Not clear how to turn this into pretty API
+
     (let [^ByteBuffer b (.buf vb)
           ^FloatBuffer fb (.asFloatBuffer b)]
-      (if (not= (alength normals-indices) 0)
-        (dotimes [vi vcount]
-          (let [pi (aget positions-indices vi)
-                ni (aget normals-indices vi)
-                ti (aget texcoords-indices vi)]
-            (.put fb positions (umul 3 pi) 3)
-            (.put fb normals (umul 3 ni) 3)
-            (.put fb texcoords (umul 2 ti) 2)))
-        (dotimes [vi vcount]
-          (let [pi (aget positions-indices vi)
-                ti (aget texcoords-indices vi)]
-            (.put fb positions (umul 3 pi) 3)
-            (.put fb 0.0)
-            (.put fb 0.0)
-            (.put fb -1.0)
-            (.put fb texcoords (umul 2 ti) 2)))))
+      (dotimes [i vcount]
+        (let [vi (aget indices i)
+              normal (float-array [0.0 0.0 -1.0])]
+          (.put fb positions (umul 3 vi) 3)
+          (if (not= (alength normals) 0)
+            (.put fb normals (umul 3 vi) 3)
+            (.put fb normal 0 3))
+          (.put fb texcoords (umul 2 vi) 2)
+          )))
```

Comment:
All this fluff in the editor is cumbersome.
I think we should pack as much data from C++ given the desired vertex declaration.

Opening a small model (~4000 vertices) takes a significant amount of time.
Loading the model in C++ takes ~0.001s, so we need to measure in the java+clojure code next. Not a new problem afaict, so we'll add a separate ticket for it.

Reply:
Yeah, we should be able to just prepare the buffers in the cpp code and use something like `NewDirectByteBuffer` to pass the data directly to the `.glBufferData` method. I'm not sure if we should be using the `VertexBuffer` record for this, but we definitely shouldn't be populating it like this. Perhaps we should just be creating a new `VertexBuffer` wrapping the `ByteBuffer` returned from cpp whenever there's an update. Not sure yet, but we shouldn't need to do much work at all on the editor side.

https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#NewDirectByteBuffer